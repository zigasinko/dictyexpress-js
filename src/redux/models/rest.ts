export type BasketAddSamplesResponse = {
    id: string;
    modified: string;
    ignored: number[];
    permitted_organisms: string[];
    permitted_sources: string[];
    conflict_organisms: string[];
    conflict_sources: string[];
};

/**
 * Process statuses.
 */
export const STATUS = {
    UPLOADING: 'UP',
    RESOLVING: 'RE',
    WAITING: 'WT',
    PREPARING: 'PP',
    PROCESSING: 'PR',
    DONE: 'OK',
    ERROR: 'ER',
    DIRTY: 'DR',
    NO_STATUS: 'NO',
};
