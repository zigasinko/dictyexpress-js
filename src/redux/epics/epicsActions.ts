import { createAction } from '@reduxjs/toolkit';
import { Gene } from 'redux/models/internal';

// Export all epic-specific actions to omit dependency-cycles for reactive queries.
// TimeSeriesEpics.
export const selectGenes = createAction<Gene[]>('genes/selectGenes');
export const pasteGenesNames = createAction<string[]>('genes/pasteGenesNames');
export const fetchTimeSeries = createAction('timeSeries/fetchTimeSeries');
export const fetchTimeSeriesSamplesExpressions = createAction(
    'timeSeries/fetchTimeSeriesSamplesExpressions',
);
export const fetchDifferentialExpressionsData = createAction(
    'timeSeries/fetchDifferentialExpressionsData',
);

// AuthenticationEpics.
export const login = createAction<{ username: string; password: string }>('authentication/login');
export const loginSucceeded = createAction('authentication/loginSucceeded');
export const logout = createAction('authentication/logout');
export const logoutSucceeded = createAction('authentication/logoutSucceeded');

// ConnectToServerEpic.
export const appStarted = createAction('appStarted');
export const connectToServer = createAction<{
    url: string;
}>('connectToServer/connect');
export const reconnectToServer = createAction('connectToServer/reconnect');
export const disconnectFromServer = createAction('connectToServer/disconnect');
export const connectionReady = createAction('connectToServer/connectionReady');

// GenesEpics.
export const fetchSelectedDifferentialExpressionGenes = createAction(
    'genes/fetchSelectedDifferentialExpressionGenes',
);
export const fetchAssociationsGenes = createAction<{ geneIds: string[]; species?: string }>(
    'genes/fetchAssociationsGenes',
);

// GOEnrichmentEpics.
export const getOrCreateGOEnrichment = createAction('gOEnrichment/getOrCreateGOEnrichment');
export const fetchGOEnrichmentData = createAction<number>('gOEnrichment/fetchGOEnrichmentData');
export const fetchGOEnrichmentStorage = createAction<number>(
    'gOEnrichment/fetchGOEnrichmentStorage',
);
