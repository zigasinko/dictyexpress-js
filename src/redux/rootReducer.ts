import timeSeries, {
    getComparisonTimeSeries,
    getSelectedTimeSeries,
    getSelectedTimeSeriesLabels,
} from 'redux/stores/timeSeries';
import genes, { getGenesById, getSelectedGenes } from 'redux/stores/genes';
import genesSimilarities, { getGenesSimilaritiesQueryGeneId } from 'redux/stores/genesSimilarities';
import samplesExpressions, {
    getSamplesExpressionsById,
    getSamplesExpressionsSamplesIds,
} from 'redux/stores/samplesExpressions';
import notifications from 'redux/stores/notifications';
import authentication from 'redux/stores/authentication';
import differentialExpressions from 'redux/stores/differentialExpressions';
import layouts from 'redux/stores/layouts';
import gOEnrichment from 'redux/stores/gOEnrichment';
import clustering from 'redux/stores/clustering';
import { combineReducers, createSelector } from '@reduxjs/toolkit';
import _ from 'lodash';
import { EMPTY_ARRAY } from 'components/genexpress/common/constants';
import { Relation } from '@genialis/resolwe/dist/api/types/rest';
import { Gene, GeneExpression, SamplesGenesExpressionsById } from './models/internal';

const rootReducer = combineReducers({
    layouts,
    authentication,
    timeSeries,
    genes,
    genesSimilarities,
    samplesExpressions,
    differentialExpressions,
    gOEnrichment,
    clustering,
    notifications,
});

export type RootState = ReturnType<typeof rootReducer>;

export type BookmarkReduxState = {
    timeSeries: Pick<RootState['timeSeries'], 'selectedId' | 'comparisonIds'>;
    genes: Pick<RootState['genes'], 'selectedGenesIds' | 'highlightedGenesIds'> & {
        source: string;
        species: string;
    };
    gOEnrichment: Pick<RootState['gOEnrichment'], 'pValueThreshold'>;
    clustering: Pick<RootState['clustering'], 'distanceMeasure' | 'linkageFunction'>;
    genesSimilarities: Pick<RootState['genesSimilarities'], 'distanceMeasure' | 'queryGeneId'>;
    differentialExpressions: Pick<RootState['differentialExpressions'], 'selectedId'>;
};

export default rootReducer;

const getTimeSeriesGenesExpressions = (
    singleTimeSeries: Relation | null,
    labels: string[],
    selectedGenes: Gene[],
    samplesExpressionsById: SamplesGenesExpressionsById,
    samplesExpressionsSamplesIds: number[],
): GeneExpression[] => {
    if (
        singleTimeSeries == null ||
        _.isEmpty(samplesExpressionsById) ||
        selectedGenes.length === 0
    ) {
        return EMPTY_ARRAY;
    }

    const newGenesExpressionsData = [] as GeneExpression[];

    labels.forEach((label) => {
        const timePointPartitions = _.flatten(
            singleTimeSeries.partitions.filter((partition) => partition.label === label),
        );

        if (
            timePointPartitions.some(
                (partition) => !samplesExpressionsSamplesIds.includes(partition.entity),
            )
        ) {
            return;
        }

        selectedGenes.forEach((gene) => {
            const values: number[] = [];
            // Gene expressions in different samples must be averaged out (mean).
            timePointPartitions.forEach((partition) => {
                if (samplesExpressionsById[partition.entity][gene.feature_id] != null) {
                    values.push(samplesExpressionsById[partition.entity][gene.feature_id]);
                }
            });

            if (values.length > 0) {
                newGenesExpressionsData.push({
                    timeSeriesName: singleTimeSeries.collection.name,
                    label,
                    value: _.mean(values),
                    geneId: gene.feature_id,
                    geneName: gene.name,
                });
            }
        });
    });

    return newGenesExpressionsData;
};

export const getSelectedGenesExpressions = createSelector(
    (state: RootState) => getSelectedTimeSeries(state.timeSeries),
    (state: RootState) => getSelectedTimeSeriesLabels(state.timeSeries),
    (state: RootState) => getSelectedGenes(state.genes),
    (state: RootState) => getSamplesExpressionsById(state.samplesExpressions),
    (state: RootState) => getSamplesExpressionsSamplesIds(state.samplesExpressions),
    getTimeSeriesGenesExpressions,
);

export const getSelectedGenesComparisonExpressions = createSelector(
    (state: RootState) => getComparisonTimeSeries(state.timeSeries),
    (state: RootState) => getSelectedTimeSeriesLabels(state.timeSeries),
    (state: RootState) => getSelectedGenes(state.genes),
    (state: RootState) => getSamplesExpressionsById(state.samplesExpressions),
    (state: RootState) => getSamplesExpressionsSamplesIds(state.samplesExpressions),
    (
        comparisonTimeSeries,
        timeSeriesLabels,
        selectedGenes,
        samplesExpressionsById,
        samplesExpressionsSamplesIds,
    ) => {
        if (comparisonTimeSeries.length === 0) {
            return [];
        }

        return comparisonTimeSeries.flatMap((comparisonSingleTimeSeries) =>
            getTimeSeriesGenesExpressions(
                comparisonSingleTimeSeries,
                timeSeriesLabels,
                selectedGenes,
                samplesExpressionsById,
                samplesExpressionsSamplesIds,
            ),
        );
    },
);

export const getGenesSimilaritiesQueryGene = createSelector(
    (state: RootState) => getGenesById(state.genes),
    (state: RootState) => getGenesSimilaritiesQueryGeneId(state.genesSimilarities),
    (genesById, genesSimilaritiesQueryGeneId) => {
        if (_.isEmpty(genesById) || genesSimilaritiesQueryGeneId == null) {
            return null;
        }

        return genesById[genesSimilaritiesQueryGeneId];
    },
);
