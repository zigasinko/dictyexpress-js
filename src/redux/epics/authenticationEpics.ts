import { ofType, Epic, combineEpics } from 'redux-observable';
import { map, startWith, endWith, catchError, mergeMap } from 'rxjs/operators';
import { of, from } from 'rxjs';
import { RootState } from 'redux/rootReducer';
import { pushToSentryAndAddErrorSnackbar } from 'redux/stores/notifications';
import {
    fetchUserSucceeded,
    loginStarted,
    fetchUserStarted,
    fetchUserEnded,
    loginEnded,
    logoutStarted,
    logoutEnded,
} from 'redux/stores/authentication';
import * as authApi from 'api/authApi';
import * as userApi from 'api/userApi';
import { clearObservers } from 'api/queryObserverManager';
import { appStarted, login, loginSucceeded, logout, logoutSucceeded } from './epicsActions';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const loginEpic: Epic<any, any, RootState, any> = (action$) =>
    action$.pipe(
        ofType(login),
        mergeMap((action) => {
            return from(authApi.login(action.payload.username, action.payload.password)).pipe(
                mergeMap(() => {
                    return from(clearObservers()).pipe(map(loginSucceeded));
                }),
                catchError((error) =>
                    of(pushToSentryAndAddErrorSnackbar(`Error logging in.`, error)),
                ),
                startWith(loginStarted()),
                endWith(loginEnded()),
            );
        }),
    );

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const logoutEpic: Epic<any, any, RootState, any> = (action$) =>
    action$.pipe(
        ofType(logout),
        mergeMap(() => {
            return from(authApi.logout()).pipe(
                mergeMap(() => {
                    return from(clearObservers()).pipe(map(logoutSucceeded));
                }),
                catchError((error) =>
                    of(pushToSentryAndAddErrorSnackbar(`Error logging out.`, error)),
                ),
                startWith(logoutStarted()),
                endWith(logoutEnded()),
            );
        }),
    );

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getCurrentUserEpic: Epic<any, any, RootState, any> = (action$) =>
    action$.pipe(
        ofType(appStarted.toString(), loginSucceeded.toString(), logoutSucceeded.toString()),
        mergeMap(() => {
            return from(userApi.getCurrentUser()).pipe(
                map(fetchUserSucceeded),
                catchError((error) =>
                    of(pushToSentryAndAddErrorSnackbar(`Error fetching user profile`, error)),
                ),
                startWith(fetchUserStarted()),
                endWith(fetchUserEnded()),
            );
        }),
    );

export default combineEpics(loginEpic, logoutEpic, getCurrentUserEpic);
