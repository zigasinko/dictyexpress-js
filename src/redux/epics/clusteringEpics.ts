import { combineLatest, of } from 'rxjs';
import { RootState } from 'redux/rootReducer';
import { getGenesById, getSelectedGenes } from 'redux/stores/genes';
import { handleError } from 'utils/errorUtils';
import _ from 'lodash';
import { getBasketExpressionsIds } from 'redux/stores/timeSeries';
import {
    clusteringDataFetchEnded,
    clusteringDataFetchStarted,
    mergedClusteringDataFetchSucceeded,
    getClusteringDistanceMeasure,
    getClusteringLinkageFunction,
    getMergedClusteringData,
} from 'redux/stores/clustering';
import {
    getSourceFromFeatures,
    getSpeciesFromFeatures,
} from '@genialis/resolwe/dist/api/types/utils';
import { GenesById, MergedClusteringData } from 'redux/models/internal';
import { Feature } from '@genialis/resolwe/dist/api/types/modules';
import { GeneClustering, Storage } from '@genialis/resolwe/dist/api/types/rest';
import { ClusteringData } from 'redux/models/rest';
import { filter, switchMap } from 'rxjs/operators';
import { fetchClusteringDataSucceeded } from './epicsActions';
import getProcessDataEpicsFactory, {
    ProcessDataEpicsFactoryProps,
    ProcessesInfo,
} from './getProcessDataEpicsFactory';
import { mapStateSlice } from './rxjsCustomFilters';

export const mergeClusteringData = (
    clustering: GeneClustering,
    genesById: GenesById,
): MergedClusteringData => {
    return {
        order: _.map(clustering.order, (nodeIndex, order) => {
            const gene = genesById[clustering.gene_symbols[nodeIndex].gene];
            return { nodeIndex, order, gene };
        }),
        linkage: _.map(clustering.linkage, ([node1, node2, distance], arrIndex) => {
            return { nodeIndex: arrIndex + clustering.order.length, node1, node2, distance };
        }),
    };
};

const processParametersObservable: ProcessDataEpicsFactoryProps<ClusteringData>['processParametersObservable'] = (
    _action$,
    state$,
) => {
    return combineLatest([
        state$.pipe(
            mapStateSlice((state) => {
                return getBasketExpressionsIds(state.timeSeries);
            }),
        ),
        state$.pipe(
            mapStateSlice((state) => {
                return getSelectedGenes(state.genes);
            }),
        ),
        state$.pipe(
            mapStateSlice((state) => {
                return getClusteringDistanceMeasure(state.clustering);
            }),
        ),
        state$.pipe(
            mapStateSlice((state) => {
                return getClusteringLinkageFunction(state.clustering);
            }),
        ),
    ]).pipe(
        filter(() => getMergedClusteringData(state$.value.clustering) == null),
        switchMap(([expressionsIds, selectedGenes, distanceMeasure, linkageFunction]) => {
            // The {Pearson/Spearman} correlation between genes must be computed on at least
            // two genes.
            if (selectedGenes.length < 2) {
                return of({});
            }

            // If basket expressions aren't in store yet, hierarchical clustering can't be
            // computed.
            if (expressionsIds.length === 0) {
                return of({});
            }

            let source;
            let species;
            try {
                source = getSourceFromFeatures(selectedGenes as Feature[]);
                species = getSpeciesFromFeatures(selectedGenes as Feature[]);
            } catch (error) {
                return of(
                    handleError(
                        `Error creating hierarchical clustering process: ${error.message}`,
                        error,
                    ),
                );
            }

            return of({
                expressions: _.sortBy(expressionsIds),
                ...(!_.isEmpty(selectedGenes) && {
                    genes: _.sortBy(_.map(selectedGenes, (gene) => gene.feature_id)),
                    gene_source: source,
                    gene_species: species,
                }),
                distance: distanceMeasure,
                linkage: linkageFunction,
                ordering: true,
            });
        }),
    );
};

const getClusteringProcessDataEpics = getProcessDataEpicsFactory<ClusteringData>({
    processInfo: ProcessesInfo.HierarchicalClustering,
    processParametersObservable,
    processStartedActionCreator: clusteringDataFetchStarted,
    processEndedActionCreator: clusteringDataFetchEnded,
    fetchDataSucceededActionCreator: fetchClusteringDataSucceeded,
    getStorageIdFromData: (data) => {
        return data.output.cluster;
    },
    actionFromStorageResponse: (storage: Storage, state: RootState) =>
        mergedClusteringDataFetchSucceeded(
            mergeClusteringData(storage.json, getGenesById(state.genes)),
        ),
});

export default getClusteringProcessDataEpics;
