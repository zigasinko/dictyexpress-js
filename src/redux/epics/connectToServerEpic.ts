import { createAction } from '@reduxjs/toolkit';
import { ofType, Epic } from 'redux-observable';
import {
    map,
    startWith,
    catchError,
    switchMap,
    takeUntil,
    retryWhen,
    delay,
    take,
} from 'rxjs/operators';
import { concat, of, throwError } from 'rxjs';
import { RootState } from 'redux/rootReducer';
import { webSocket } from 'rxjs/webSocket';
import { handleWebSocketMessage, sessionId } from 'api/queryObserverManager';
import { Message } from '@genialis/resolwe/dist/api/connection';
import { pushToSentryAndAddErrorSnackbar } from 'redux/stores/notifications';

// ReconnectionTimeout = 60s.
const reconnectionTimeout = 6000;
const reconnectionMaxAttempts = 3;

// Export epic actions.
export const appStarted = createAction('appStarted');
export const connectToServer = createAction<{
    url: string;
}>('connectToServer/connect');
export const reconnectToServer = createAction('connectToServer/reconnect');
export const disconnectFromServer = createAction('connectToServer/disconnect');
export const connectionReady = createAction('connectToServer/connectionReady');

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const connectToWebSocketServiceEpic: Epic<any, any, RootState, any> = (action$) =>
    action$.pipe(
        ofType(appStarted),
        map(() => {
            return connectToServer({
                url: 'wss://qa2.genialis.com/ws/',
            });
        }),
    );

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const connectToServerEpic: Epic<any, any, RootState, any> = (action$) =>
    action$.pipe(
        ofType(connectToServer),
        switchMap(({ payload: { url } }) => {
            return action$.pipe(
                ofType(reconnectToServer),
                takeUntil(action$.pipe(ofType(disconnectFromServer))),
                startWith(null),
                map(() => {
                    return webSocket({
                        url: `${url + sessionId}?subscribe-broadcast`,
                        WebSocketCtor: WebSocket,
                    });
                }),
                switchMap((webSocketConnection$) =>
                    webSocketConnection$.pipe(
                        takeUntil(
                            action$.pipe(
                                ofType(
                                    reconnectToServer.toString(),
                                    disconnectFromServer.toString(),
                                ),
                            ),
                        ),
                        map((message) => {
                            return handleWebSocketMessage(message as Message);
                        }),
                        retryWhen((errors) =>
                            errors.pipe(
                                delay(reconnectionTimeout),
                                take(reconnectionMaxAttempts),
                                (o) => concat(o, throwError(o)),
                            ),
                        ),
                        catchError((err) => {
                            return of(
                                pushToSentryAndAddErrorSnackbar(
                                    `WebSocket connection failed after ${reconnectionMaxAttempts} tries.`,
                                    err,
                                ),
                            );
                        }),
                        startWith(connectionReady()),
                    ),
                ),
            );
        }),
    );
