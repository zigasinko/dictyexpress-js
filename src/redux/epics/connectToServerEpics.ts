import { Action } from '@reduxjs/toolkit';
import { ofType, Epic } from 'redux-observable';
import { startWith, catchError, takeUntil, retryWhen, delay, take, mergeMap } from 'rxjs/operators';
import { merge, of, throwError } from 'rxjs';
import { RootState } from 'redux/rootReducer';
import { webSocket } from 'rxjs/webSocket';
import { Message } from '@genialis/resolwe/dist/api/connection';
import { handleError } from 'utils/errorUtils';
import { sessionId, webSocketUrl } from 'api/base';
import { handleWebSocketMessage } from 'managers/queryObserverManager';
import { filterNullAndUndefined } from './rxjsCustomFilters';
import {
    appStarted,
    connectionReady,
    disconnectFromServer,
    reconnectToServer,
} from './epicsActions';

const reconnectionTimeout = 6000;
const reconnectionMaxAttempts = 3;

const connectToServerEpic: Epic<Action, Action, RootState> = (action$) =>
    action$.pipe(
        ofType(appStarted.toString(), reconnectToServer.toString()),
        takeUntil(action$.pipe(ofType(disconnectFromServer))),
        mergeMap(() =>
            webSocket({
                url: `${webSocketUrl}/${sessionId}`,
                WebSocketCtor: WebSocket,
            }).pipe(
                mergeMap((message) => handleWebSocketMessage(message as Message)),
                retryWhen((errors) =>
                    errors.pipe(delay(reconnectionTimeout), take(reconnectionMaxAttempts), (o) =>
                        merge(o, throwError(o)),
                    ),
                ),
                catchError((err) =>
                    of(
                        handleError(
                            `WebSocket connection failed after ${reconnectionMaxAttempts} tries.`,
                            err,
                        ),
                    ),
                ),
                takeUntil(
                    action$.pipe(
                        ofType(reconnectToServer.toString(), disconnectFromServer.toString()),
                    ),
                ),
                startWith(connectionReady()),
            ),
        ),
        filterNullAndUndefined(),
    );

export default connectToServerEpic;
