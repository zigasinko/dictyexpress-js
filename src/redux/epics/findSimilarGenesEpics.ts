import { RootState } from 'redux/rootReducer';
import { getSelectedGenesIds } from 'redux/stores/genes';
import _ from 'lodash';
import { getBasketExpressionsIds } from 'redux/stores/timeSeries';
import { Storage } from '@genialis/resolwe/dist/api/types/rest';
import { FindSimilarGenesData } from 'redux/models/rest';
import {
    getGenesSimilaritiesDistanceMeasure,
    getGenesSimilarities,
    getGenesSimilaritiesQueryGeneId,
    genesSimilaritiesFetchEnded,
    genesSimilaritiesFetchStarted,
    genesSimilaritiesFetchSucceeded,
    genesSimilaritiesQueryGeneSet,
} from 'redux/stores/genesSimilarities';
import { combineEpics, Epic, ofType } from 'redux-observable';
import { filter, map, switchMap } from 'rxjs/operators';
import { Action } from '@reduxjs/toolkit';
import { combineLatest, of } from 'rxjs';
import { fetchGenesSimilarities, fetchGenesSimilaritiesDataSucceeded } from './epicsActions';
import getProcessDataEpicsFactory, {
    ProcessDataEpicsFactoryProps,
    ProcessesInfo,
} from './getProcessDataEpicsFactory';
import { mapStateSlice } from './rxjsCustomFilters';

const processParametersObservable: ProcessDataEpicsFactoryProps<FindSimilarGenesData>['processParametersObservable'] = (
    action$,
    state$,
) => {
    return combineLatest([
        action$.pipe(ofType(fetchGenesSimilarities)),
        state$.pipe(
            mapStateSlice((state) => {
                return getBasketExpressionsIds(state.timeSeries);
            }),
        ),
        state$.pipe(
            mapStateSlice((state) => {
                return getGenesSimilaritiesQueryGeneId(state.genesSimilarities) ?? '';
            }),
        ),
        state$.pipe(
            mapStateSlice((state) => {
                return getGenesSimilaritiesDistanceMeasure(state.genesSimilarities);
            }),
        ),
    ]).pipe(
        filter(() => getGenesSimilarities(state$.value.genesSimilarities).length === 0),
        switchMap(
            // Unused _fetchGenesSimilarities var is necessary to keep rxjs from piping before
            // fetchGenesSimilarities action is emitted (after find similar genes modal is opened).
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            ([_fetchGenesSimilarities, expressionsIds, queryGeneId, distanceMeasure]) => {
                // The {Pearson/Spearman} correlation between genes must be computed on at least
                // two genes.
                if (queryGeneId === '') {
                    return of({});
                }

                // If basket expressions aren't in store yet, hierarchical clustering can't be
                // computed.
                if (expressionsIds.length === 0) {
                    return of({});
                }

                return of({
                    expressions: _.sortBy(expressionsIds),
                    gene: queryGeneId,
                    distance: distanceMeasure,
                });
            },
        ),
    );
};

const getFindSimilarGenesProcessDataEpics = getProcessDataEpicsFactory<FindSimilarGenesData>({
    processInfo: ProcessesInfo.FindSimilarGenes,
    processParametersObservable,
    processStartedActionCreator: genesSimilaritiesFetchStarted,
    processEndedActionCreator: genesSimilaritiesFetchEnded,
    fetchDataSucceededActionCreator: fetchGenesSimilaritiesDataSucceeded,
    getStorageIdFromData: (data) => {
        return data.output.similar_genes;
    },
    actionFromStorageResponse: (storage: Storage) =>
        genesSimilaritiesFetchSucceeded(storage.json['similar genes']),
});

const handleSelectedGenesChangedEpic: Epic<Action, Action, RootState> = (action$, state$) => {
    return state$.pipe(
        mapStateSlice(
            (state) => getSelectedGenesIds(state.genes),
            (selectedGenesIds) => selectedGenesIds.length > 0,
        ),
        filter((selectedGenesIds) => {
            const queryGeneId = getGenesSimilaritiesQueryGeneId(state$.value.genesSimilarities);
            return queryGeneId == null || !selectedGenesIds.includes(queryGeneId);
        }),
        map((selectedGenesIds) => genesSimilaritiesQueryGeneSet(selectedGenesIds[0] ?? null)),
    );
};

export default combineEpics(getFindSimilarGenesProcessDataEpics, handleSelectedGenesChangedEpic);
