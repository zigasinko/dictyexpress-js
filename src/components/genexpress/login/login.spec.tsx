import React from 'react';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import { customRender } from 'tests/test-utils';
import { testState, mockStore } from 'tests/mock';
import { MockStoreEnhanced } from 'redux-mock-store';
import { RootState } from 'redux/rootReducer';
import { AppDispatch } from 'redux/appStore';
import { User } from '@genialis/resolwe/dist/api/types/rest';
import { login } from 'redux/epics/authenticationEpics';
import Login from './login';

const initialTestState = testState();

it('should display sign in form', () => {
    const { asFragment } = customRender(<Login isLoginModalOpened closeModal={(): void => {}} />);

    expect(asFragment()).toMatchSnapshot();
});

describe('modal opened', () => {
    let mockedStore: MockStoreEnhanced<RootState, AppDispatch>;
    let asFragment: () => DocumentFragment;

    const usernameValue = 'asdf';
    const passwordValue = 'asdf';

    beforeEach(() => {
        initialTestState.authentication.user = {} as User;
        initialTestState.authentication.isLoggedIn = false;
        mockedStore = mockStore(initialTestState);
        mockedStore.clearActions();

        ({ asFragment } = customRender(<Login isLoginModalOpened closeModal={(): void => {}} />, {
            mockedStore,
        }));
    });

    it('should display sign in form', () => {
        expect(asFragment()).toMatchSnapshot();
    });

    it('should mask password value', () => {
        fireEvent.change(screen.getByLabelText('Password'), {
            target: { value: passwordValue },
        });

        expect(screen.getByDisplayValue(passwordValue)).toHaveAttribute('type', 'password');
    });

    it('should disable submit button until username and password are entered', () => {
        expect(screen.getByRole('button', { name: 'SIGN IN' })).toBeDisabled();

        fireEvent.change(screen.getByLabelText('Username'), {
            target: { value: usernameValue },
        });

        fireEvent.change(screen.getByLabelText('Password'), {
            target: { value: passwordValue },
        });

        expect(screen.getByRole('button', { name: 'SIGN IN' })).toBeEnabled();
    });

    it('should trigger login action after user clicks SIGN IN', async () => {
        fireEvent.change(screen.getByLabelText('Username'), {
            target: { value: usernameValue },
        });

        fireEvent.change(screen.getByLabelText('Password'), {
            target: { value: passwordValue },
        });

        fireEvent.click(await screen.findByText('SIGN IN'));

        await waitFor(() =>
            expect(mockedStore.getActions()).toEqual([
                login({ username: usernameValue, password: passwordValue }),
            ]),
        );
    });
});
