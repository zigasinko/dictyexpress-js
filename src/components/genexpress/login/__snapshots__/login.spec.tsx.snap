// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal opened should display sign in form 1`] = `<DocumentFragment />`;

exports[`should display sign in form 1`] = `<DocumentFragment />`;
